File: stringcalc/javascript/test/stringCalcSpec.js
var should = require('should');
var stringCalc = require('./../stringCalc.js');

describe("stringCalc", function() {
	
	it("It is an object that exists", function(){
		should.exist(stringCalc);
		stringCalc.should.be.a("object");
	});

	describe("stringCalc.add",function(){
		it("should be a function", function(){
			should.exist(stringCalc.add);
		});

		describe("when passed an empty string", function(){
			it("should return 0", function(){
				(stringCalc.add('')).should.equal(0);	
			});
		});

		describe("when passed in a single value", function(){
			it("should return that value", function(){
				(stringCalc.add('1')).should.equal(1);
			});
		});

		describe("when passed in two values", function(){
			it("should return the sum of those values", function(){
				(stringCalc.add('1,2')).should.equal(3);
			});
		});

		describe("when calling Add with a negative number will throw an exception", function() {
			it("should throw an exception because of a negative number", function() {
				(function() {
					stringCalc.add('1,-1');
				}).should.throw();
			});
		});

	});
});

File: stringcalc/ruby/test/string_calc_spec.rb
require 'minitest/spec'
require 'minitest/autorun'
require 'minitest/reporters'
MiniTest::Reporters.use!
require_relative '../string_calc'

describe 'StringCalc' do

  describe 'class methods' do

    describe '#add' do

      describe 'when passed an empty string' do

        it 'returns zero' do
          StringCalc.add('').must_equal 0
        end

      end

      describe 'when passed a single value' do
        it 'returns that value' do
          StringCalc.add('3').must_equal 3
        end
      end

      describe 'when passed two, comma-delimited values' do
        it 'returns the sum of those two values' do
          StringCalc.add('3,4').must_equal 7
        end
      end

      describe 'when passed three, comma-delimited values' do
        it 'returns the sum of those values' do
          StringCalc.add('3,4,5').must_equal 12
        end
      end

      describe 'when a newline character (\n) is added between values' do
        it 'is ignored and the values are summed as if it were not there' do
          StringCalc.add("1, 2, \n 3").must_equal 6
        end
      end

      describe 'when passing a comma-delimited string of values that contain negative integers' do
        it 'throws an exception' do
          lambda {
            StringCalc.add('1,-2,4')
          }.must_raise StringCalc::ArithmeticError
        end

        it 'includes a list of the negative values in the message from the exception' do
          exception = lambda {
            StringCalc.add('1,-2,4')
          }.must_raise StringCalc::ArithmeticError

          exception.message.must_equal 'Values cannot include: -2'
        end
      end

      describe 'when passed a custom delimiter as the second argument' do
        it 'uses that delimiter to split the strings and sum the values' do
          StringCalc.add('1#2#3', '#').must_equal 6
        end
      end

      describe 'when passed invalid input (e.g. alpha/symbol)' do
        it 'returns zero' do
          StringCalc.add('1,a,$,2').must_equal 0
        end
      end

    end

  end

end